lib/pdk/tests/unit.rb

```diff
@@ -6,20 +6,96 @@ module PDK
   module Test
     class Unit
       def self.cmd(_tests)
-        # TODO: actually run the tests
-        # cmd = 'rake spec'
-        # cmd += " #{tests}" if tests
-        cmd = 'pwd'
-        cmd
+        # TODO: test selection
+        [File.join(PDK::Util.module_root, 'bin', 'rake'), 'spec']
       end
 
-      def self.invoke(tests, report = nil)
+      def self.invoke(report, options = {})
         PDK::Util::Bundler.ensure_bundle!
+        PDK::Util::Bundler.ensure_binstubs!('rake')
 
-        puts _('Running unit tests: %{tests}') % { tests: tests }
+        tests = options.fetch(:tests)
 
-        output = PDK::CLI::Exec.execute(cmd(tests))
-        report.write(output) if report
+        cmd_argv = cmd(tests)
+        cmd_argv.unshift('ruby') if Gem.win_platform?
+
+        command = PDK::CLI::Exec::Command.new(*cmd_argv).tap do |c|
+          c.context = :module
+          c.add_spinner('Running unit tests')
+          c.environment['CI_SPEC_OPTIONS'] = '--format j'
+        end
+
+        PDK.logger.debug(_('Running %{cmd}') % { cmd: command.argv.join(' ') })
+
+        result = command.execute!
+
+        # TODO: cleanup rspec and/or beaker output
+        # Iterate through possible JSON documents until we find one that is valid.
+        json_result = nil
+
+        result[:stdout].scan(%r{\{(?:[^{}]|(?:\g<0>))*\}}x) do |str|
+          begin
+            json_result = JSON.parse(str)
+            break
+          rescue JSON::ParserError
+            next
+          end
+        end
+
+        raise PDK::CLI::FatalError, _('Unit test output did not contain a valid JSON result: %{output}') % { output: result[:stdout] } unless json_result
+
+        parse_output(report, json_result)
+
+        result[:exit_code]
+      end
+
+      def self.parse_output(report, json_data)
+        # Output messages to stderr.
+        json_data['messages'] && json_data['messages'].each { |msg| $stderr.puts msg }
+
+        example_results = {
+          # Only possibilities are passed, failed, pending:
+          # https://github.com/rspec/rspec-core/blob/master/lib/rspec/core/example.rb#L548
+          'passed' => [],
+          'failed' => [],
+          'pending' => [],
+        }
+
+        json_data['examples'] && json_data['examples'].each do |ex|
+          example_results[ex['status']] << ex if example_results.key?(ex['status'])
+        end
+
+        example_results.each do |result, examples|
+          # Translate rspec example results to JUnit XML testcase results
+          state = case result
```

Comment:
FYI I created https://github.com/puppetlabs/pdk/pull/115 (now merged) to support pending/skipped results better